[ML] Remove dependency cache.

walterra opened the pull request:
## Summary

Fixes #153477.
Fixes #153476.
Part of #187772 (technical debt).
Part of #153288 (migrate enzyme tests to react-testing-lib).

Removes dependency cache. The major culprit making this PR large and not easy to split is that `getHttp()` from the dependency cache was used throughout the code base for services like `mlJobService` and `ml/mlApiServices` which then themselves were directly imported and not part of React component lifecycles. 

- For functional components this means mostly migrating to hooks that allow accessing services.
- We still have a bit of a mix of usage of `withKibana` and `context` for class based React components. This was not consolidated in this PR, I took what's there and adjusted how services get used. These components access services via `this.props.kibana.services.*` or `this.context.services.*`.
- Functions no longer access the global services provided via dependency cache but were updated to receive services via arguments.
- Stateful services like `mlJobService` are exposed now via a factory that makes sure the service gets instantiated only once.
- Some tests where the mocks needed quite some refactoring were ported to `react-testing-lib`. They no longer make use of snapshots or call component methods which should be considered implementation details. 
- We have a mix of usage of the plain `toasts` via `useMlKibana` and our own `toastNotificationServices` that wraps `toasts`. I didn't consolidate this in this PR but used what's available for the given code.
- For class based components, service initializations were moved from `componentDidMount()` to `constructor()` where I spotted it.
- We have a bit of a mix of naming: `ml`, `mlApiServices`, `useMlApiContext()` for the same thing. I didn't consolidate the naming in this PR, to avoid making this PR even larger. This can be done in a follow up, once this PR is in this should be more straightforward and less risky. 
- Turns out `explorer_chart_config_builder.js` is no longer used anywhere so I deleted it.
- `jobs/jobs_list/components/utils.d.ts` was missing some definitions, tried to fix them.
- Moved `stashJobForCloning` to be a method of `mlJobService`.
- The `MetricSelector` component was an exact copy besides the i18n label, consolidated that so anomaly detection wizards use the same component.

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

walterra commented:
/ci

walterra commented:
/ci

walterra commented:
/ci

walterra commented:
/ci

walterra commented:
/ci

walterra commented:
/ci

walterra commented:
/ci

walterra commented:
/ci

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

darnautov commented:
@walterra Single Metric Viewer should not display a chart until the partition field is selected. 

<img width="1756" alt="image" src="https://github.com/user-attachments/assets/f5e850dd-fb49-49ec-8fd1-0c24108c784e">

On `main` it looks like this: 

<img width="1761" alt="image" src="https://github.com/user-attachments/assets/95b2fc8d-5973-4dcf-a38b-4be2f49b4642">

darnautov commented:
When I click "View" on one of the anomaly charts in the Anomaly Explorer, it doesn't plot it in the SMV. 
<img width="658" alt="image" src="https://github.com/user-attachments/assets/9ea27f7d-9345-4a01-8f4a-788c5b2203cb">

<img width="2257" alt="image" src="https://github.com/user-attachments/assets/e56dc9a2-489a-42c8-86d2-91a1ea9a184c">

darnautov commented:
After clicking "Refresh" once, up to 3 identical ongoing requests happen, e.g. `/annotations`, `/anomaly_search`
<img width="1785" alt="image" src="https://github.com/user-attachments/assets/c8e0a297-3687-4bd3-9f3d-f4144ac32431">

darnautov commented:
"Stop datafeed" action doesn't work
<img width="1561" alt="image" src="https://github.com/user-attachments/assets/18f6caad-1859-44ff-9884-e5bc975fa54c">

Console error: 
```Uncaught TypeError: jobs is undefined
    stopDatafeeds utils.js:74
    onClick management.js:53
    onClick collapsed_item_actions.js:68
    React 14
    unstable_runWithPriority scheduler.development.js:468
    React 4
        runWithPriority$1
        discreteUpdates$1
        discreteUpdates
        dispatchDiscreteEvent
```

darnautov commented:
I assume it's all related to the same issue, try this URL of the SMV
```
http://localhost:5601/abc/app/ml/timeseriesexplorer?_g=%28ml%3A%28jobIds%3A%21%28cw_multi_1%29%29%2CrefreshInterval%3A%28display%3AOff%2Cpause%3A%21t%2Cvalue%3A0%29%2Ctime%3A%28from%3A%272019-11-04T00%3A00%3A00.000Z%27%2Cmode%3Aabsolute%2Cto%3A%272024-08-14T13%3A00%3A00.000Z%27%29%29&_a=%28mlSelectInterval%3A%28display%3A%271%20day%27%2Cval%3Aday%29%2Ctimeseriesexplorer%3A%28mlTimeSeriesExplorer%3A%28detectorIndex%3A1%2Centities%3A%28instance%3Ai-70dd1dc1%29%2Czoom%3A%28from%3A%272019-11-04T11%3A30%3A00.000Z%27%2Cto%3A%272019-11-05T12%3A30%3A00.000Z%27%29%29%29%29
```
`main`:
<img width="1554" alt="image" src="https://github.com/user-attachments/assets/29ee7d1f-6ba2-43d6-af9d-a4b06f0f921e">

this PR: 
<img width="1548" alt="image" src="https://github.com/user-attachments/assets/0427b824-7ca2-45db-80c0-a53992668f36">

peteharverson commented:
The Single Metric Viewer doesn't seem to be filtering the chart or table by the selected partition field (here the chart and table are showing anomalies for values of method other than `HEAD`)

<img width="1358" alt="Screenshot 2024-08-21 at 11 32 09" src="https://github.com/user-attachments/assets/dd7552b2-69d4-429a-8794-bfd5de69ca09">

peteharverson commented:
Gave the following functionality a test:
- Embeddables in Dashboard and a case (swim lane, anomaly chart, SMV, change point)
- Overview, Notifications, Memory Usage pages
- Single Metric Viewer

Only new issue I found was the use of the partition field in the Single Metric Viewer - https://github.com/elastic/kibana/pull/189729#issuecomment-2301715849

walterra commented:
@peteharverson Looks like I missed updating some JS code to pass around the selected job properly, fixed in 977b9bf.

walterra commented:
@darnautov I looked into your other comments:

> Single Metric Viewer should not display a chart until the partition field is selected. 

This should be fixed, I also added assertions to the functional tests to make sure the charts and anomaly table are not rendered before a correct selection.

> When I click "View" on one of the anomaly charts in the Anomaly Explorer, it doesn't plot it in the SMV.

Please check if this is now fixed for you too, after the above fix I wasn't able to reproduce anymore.

> After clicking "Refresh" once, up to 3 identical ongoing requests happen, e.g. /annotations, /anomaly_search

Investigated this and I think it's the same behavior on `main`. 

> "Stop datafeed" action doesn't work

This should work now, also found and fixed another issue related to that where the toasts where not correctly rendered.

alvarezmelissa87 commented on x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_exploration/components/regression_exploration/evaluate_panel.tsx:
```diff
@@ -65,6 +65,7 @@ export const EvaluatePanel: FC<Props> = ({ jobConfig, jobStatus, searchQuery })
   const {
     services: { docLinks },
   } = useMlKibana();
+  const mlApiServices = useMlApiContext();
```
I believe `mlApiServices` is included in `services` provided by `useMlKibana` as part of `mlServices` so could do something like:

```
 const {
    services: { docLinks, mlServices: { mlApiServices } },
  } = useMlKibana();
  ```

walterra replied:
Yes, it's part of the kibana context, although it's quite deeply nested and with the prettier formatting it quite often ends up as multiple lines using the destructuring. I'd consider the hook a "shortcut" of sorts. On `main` there's also quite some similar usage where we use `useMlKibana()` to get some general services and more specific hooks like `useMlApiContext()` or `useToastNoticiations()` in the same place so I'd prefer to leave this as is.

alvarezmelissa87 commented on x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_exploration/components/exploration_results_table/use_exploration_results.ts:
```diff
@@ -45,10 +44,14 @@ import { useExplorationDataGrid } from './use_exploration_data_grid';
+      notifications: { toasts },
+    },
+  } = useMlKibana();
+  const ml = useMlApiContext();
```
Same here - not sure we need to use `useMlApiContext` since it should be included in `useMlKibana`

alvarezmelissa87 commented on x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_exploration/components/outlier_exploration/use_outlier_data.ts:
```diff
@@ -45,6 +45,12 @@ export const useOutlierData = (
+      notifications: { toasts },
+    },
+  } = useMlKibana();
+  const ml = useMlApiContext();
```
Same here - not sure we need to bring in `useMlApiContext`

alvarezmelissa87 commented on x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_management/components/action_edit/edit_action_flyout.tsx:
```diff
@@ -70,10 +69,10 @@ export const EditActionFlyout: FC<Required<EditAction>> = ({ closeFlyout, item }
   const { refresh } = useRefreshAnalyticsList();
 
-  const mlApiServices = useMlApiContext();
+  const ml = useMlApiContext();
```
Same here - I think we can use the services provided in `useMlKibana`

alvarezmelissa87 commented on x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_management/hooks/use_create_analytics_form/use_create_analytics_form.ts:
```diff
@@ -50,6 +49,7 @@ export const useCreateAnalyticsForm = (): CreateAnalyticsFormProps => {
       data: { dataViews },
     },
   } = useMlKibana();
+  const ml = useMlApiContext();
```
Same here.

alvarezmelissa87 commented on x-pack/plugins/ml/public/application/datavisualizer/file_based/file_datavisualizer.tsx:
```diff
@@ -36,8 +36,9 @@ export const FileDataVisualizerPage: FC = () => {
       },
     },
   } = useMlKibana();
+  const mlApiServices = useMlApiContext();
```
Same here

alvarezmelissa87 commented on x-pack/plugins/ml/public/application/data_frame_analytics/common/use_results_view_config.ts:
```diff
@@ -35,8 +34,11 @@ export const useResultsViewConfig = (jobId: string) => {
       data: { dataViews },
     },
   } = useMlKibana();
+  const toastNotificationService = useToastNotificationService();
```
Same here.

alvarezmelissa87 commented on x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_creation/components/advanced_step/advanced_step_form.tsx:
```diff
@@ -134,6 +134,7 @@ export const AdvancedStepForm: FC<CreateAnalyticsStepProps> = ({
   const {
     services: { docLinks },
   } = useMlKibana();
+  const mlApiServices = useMlApiContext();
```
Same here

alvarezmelissa87 commented on x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_creation/components/configuration_step/configuration_step_form.tsx:
```diff
@@ -115,6 +114,10 @@ export const ConfigurationStepForm: FC<ConfigurationStepProps> = ({
 }) => {
+  const { services } = useMlKibana();
+  const toastNotifications = services.notifications.toasts;
+  const mlApiServices = useMlApiContext();
```
Same here

alvarezmelissa87 commented on x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_creation/components/create_step_footer/create_step_footer.tsx:
```diff
@@ -49,6 +48,7 @@ export const CreateStepFooter: FC<Props> = ({ jobId, jobType, showProgress }) =>
   const {
     services: { notifications },
   } = useMlKibana();
+  const ml = useMlApiContext();
```
Same here

alvarezmelissa87 commented on x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_creation/components/details_step/details_step_form.tsx:
```diff
@@ -43,6 +42,7 @@ export const DetailsStepForm: FC<CreateAnalyticsStepProps> = ({
   const {
     services: { docLinks, notifications },
   } = useMlKibana();
+  const ml = useMlApiContext();
```
Same here

alvarezmelissa87 commented:
Awesome work getting rid of the dependency cache finally! Functionality for DFA looks good - just added a few comments to the code.

peteharverson reviewed (approved):
LGTM after latest edits for the areas I tested:
- Embeddables
- Single Metric Viewer
- Overview page
- Notifications
- Memory Usage

alvarezmelissa87 reviewed (approved):
LGTM ⚡

jgowdyelastic commented on x-pack/plugins/ml/public/application/jobs/new_job/common/job_creator/multi_metric_job_creator.ts:
```diff
@@ -90,7 +100,13 @@ export class MultiMetricJobCreator extends JobCreator {
     this.createdBy = CREATED_BY_LABEL.MULTI_METRIC;
     this._sparseData = isSparseDataJob(job, datafeed);
-    const detectors = getRichDetectors(job, datafeed, this.additionalFields, false);
+    const detectors = getRichDetectors(
```
I think `getRichDetectors` (and `getFieldByIdFactory`) could be moved to be a member of the base class. It then wouldn't need `newJobCapsService` to be passed to it.

I can do this in a follow up, if you don't want to complicate this PR any further.

walterra replied:
Yes, let's do this in follow ups please 😅 .

jgowdyelastic commented:
This a great bit of refactoring, thanks for doing it.

I'm going to follow it up with some further changes to move `stashJobForCloning` out of `mlJobService`. I think having it inside there complicates things. mlJobService's dependencies are unnecessary for the stashing of the job and moving it into it's own file or service will clean up all the places where we need to stash a job for cloning.
I don't think we need to make these changes in this PR as it'll be good to just get this merged.

walterra commented on x-pack/plugins/ml/public/embeddables/job_creation/aiops/flyout/create_job.tsx:
```diff
@@ -91,10 +96,11 @@ export const CreateJob: FC<Props> = ({ dataView, field, query, timeRange }) => {
       ),
-
-    [dashboardService, data, mlApiServices, uiSettings]
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Updated in [393abda](https://github.com/elastic/kibana/pull/189729/commits/393abda746879aae70089ee451e2e74e8c5eef36).

walterra commented on x-pack/plugins/ml/public/application/services/http_service.ts:
```diff
@@ -48,25 +47,14 @@ function getFetchOptions(options: HttpFetchOptionsWithPath): {
 
-  constructor(private httpStart: HttpStart) {
-    this.getLoadingCount$ = httpStart.getLoadingCount$();
+  constructor(private httpSetup: HttpSetup) {
```
Updated in Updated in [393abda](https://github.com/elastic/kibana/pull/189729/commits/393abda746879aae70089ee451e2e74e8c5eef36).

jgowdyelastic reviewed (approved):
LGTM

darnautov reviewed (approved):
LGTM
